toolchain/check/convert.cpp

```diff
@@ -1208,19 +1208,60 @@ auto Convert(Context& context, SemIR::LocId loc_id, SemIR::InstId expr_id,
       if (!target.diagnose) {
         return context.emitter().BuildSuppressed();
       }
-      // TODO: Should this message change to say "object of type" when
-      // converting from a reference expression?
-      CARBON_DIAGNOSTIC(ImplicitAsConversionFailure, Error,
-                        "cannot implicitly convert value of type {0} to {1}",
-                        TypeOfInstId, SemIR::TypeId);
-      CARBON_DIAGNOSTIC(ExplicitAsConversionFailure, Error,
-                        "cannot convert value of type {0} to {1} with `as`",
-                        TypeOfInstId, SemIR::TypeId);
-      return context.emitter().Build(loc_id,
-                                     target.kind == ConversionTarget::ExplicitAs
-                                         ? ExplicitAsConversionFailure
-                                         : ImplicitAsConversionFailure,
-                                     expr_id, target.type_id);
+      bool explicit_as = target.kind == ConversionTarget::ExplicitAs;
+      auto type_of_expr_id = context.insts().Get(expr_id).type_id();
+      if (target.type_id == SemIR::TypeType::SingletonTypeId ||
+          sem_ir.types().Is<SemIR::FacetType>(target.type_id)) {
+        if (context.types().Is<SemIR::FacetType>(type_of_expr_id)) {
+          CARBON_DIAGNOSTIC(
+              ConversionFailureFacetToFacet, Error,
+              "cannot{0:| implicitly} convert type {1} that implements {2} to "
```

Comment:
Usually when we talk about "convert[ing] X to Y", X describes a source expression, and Y describes a target type, so we're talking about something like `X as Y` or `let v: Y = X;`. But here, "type implementing <facet>" is a type, but it's not the target type of the conversion: converting an expression `X` to type `T` where `T:! Y` would be a "conversion to a type (T) implementing Y". So I find this phrasing inconsistent. At the same time, I do like that we say "type implementing Y" in this diagnostic, because I think that gives the reader a good clue about what's going on.

I suggest that we change the word "to" to "into" in these diagnostics, so that we use "convert X to Y" to mean a conversion of source expression X to target type Y, and we use "convert X into Y" to mean a conversion of source expression X to a target expression described by Y. So for example "convert expression X to type Y" and "convert expression X into expression of type Y" would mean the same thing.

tl;dr:
```suggestion
              "cannot{0:| implicitly} convert type {1} that implements {2} into "
```
here and in the next 3 diagnostics.

Reply:
Thanks, I also struggled with this. I like the direction here.